File: src/main/java/com/example/application/backend/AuthServlet.java
package com.example.application.backend;

import com.vaadin.flow.server.*;

import javax.servlet.ServletException;
import java.io.IOException;
import java.util.Arrays;

//@WebServlet(urlPatterns = "/*", asyncSupported = true)
public class AuthServlet extends VaadinServlet {

    @Override
    protected void servletInitialized() throws ServletException {
        super.servletInitialized();
//        getService().addSessionInitListener(e -> e.getSession().addRequestHandler(new AuthRequestHandler()));
    }

    protected class AuthRequestHandler extends SynchronizedRequestHandler {

        @Override
        public boolean synchronizedHandleRequest(VaadinSession session, VaadinRequest request, VaadinResponse response) throws IOException {
            if (Arrays.asList("/form", "/", "/masterdetail").contains(((VaadinServletRequest) request).getRequestURI())) {
                String user = (String) session.getAttribute("user");
                String password = (String) session.getAttribute("password");

                if (!AuthService.INSTANCE.authenticate(user, password)) {
                    ((VaadinServletResponse) response).getHttpServletResponse().sendRedirect("/login");
                    return true;
                }
            }

            return false;
        }
    }
}

File: src/main/kotlin/com/example/application/backend/AuthServlet.kt
package com.example.application.backend

import com.vaadin.flow.server.*
import java.io.IOException
import java.util.*
import javax.servlet.ServletException

//@WebServlet(urlPatterns = "/*", asyncSupported = true)
class AuthServlet : VaadinServlet() {

    @Throws(ServletException::class)
    override fun servletInitialized() {
        super.servletInitialized()
        //        getService().addSessionInitListener(e -> e.getSession().addRequestHandler(new AuthRequestHandler()));
    }

    protected inner class AuthRequestHandler : SynchronizedRequestHandler() {

        @Throws(IOException::class)
        override fun synchronizedHandleRequest(session: VaadinSession, request: VaadinRequest, response: VaadinResponse): Boolean {
            if (Arrays.asList("/form", "/", "/masterdetail").contains((request as VaadinServletRequest).requestURI)) {
                val user = session.getAttribute("user") as String
                val password = session.getAttribute("password") as String

                if (!AuthService.INSTANCE.authenticate(user, password)) {
                    (response as VaadinServletResponse).httpServletResponse.sendRedirect("/login")
                    return true
                }
            }

            return false
        }
    }
}
